module/zfs/vdev.c

```diff
@@ -4425,15 +4425,22 @@ vdev_set_state(vdev_t *vd, boolean_t isopen, vdev_state_t state, vdev_aux_t aux)
 	if (state == vd->vdev_state) {
 		/*
 		 * Since vdev_offline() code path is already in an offline
-		 * state we can miss a statechange event to OFFLINE. Check
-		 * the previous state to catch this condition.
+		 * state, we can miss a statechange event to OFFLINE. We can
+		 * also miss a statechange event to ONLINE when a drive is
+		 * replaced. Check the previous state to catch either
+		 * condition.
 		 */
 		if (vd->vdev_ops->vdev_op_leaf &&
 		    (state == VDEV_STATE_OFFLINE) &&
 		    (vd->vdev_prevstate >= VDEV_STATE_FAULTED)) {
 			/* post an offline state change */
 			zfs_post_state_change(spa, vd, vd->vdev_prevstate);
+		} else if (state == VDEV_STATE_ONLINE &&
+		    vd->vdev_prevstate == VDEV_STATE_UNKNOWN) {
+			/* post an online state change */
+			zfs_post_state_change(spa, vd, vd->vdev_prevstate);
 		}
```

Comment:
Good thought, but unfortunately this doesn't work quite as intended.  In my manual testing this will result in "ONLINE" `resource.fs.zfs.statechange` events being generated for both the top-level mirror/raidz vdev and the _old replaced_ leaf vdev guid (not the new vdev guid).

For this to behave as intended we'd need to add logic to the end of `vdev_set_state()`, and somehow not filter out the initial open event from a replaced vdev.

```
                /* filter out state change due to initial vdev_open */
                if (save_state > VDEV_STATE_CLOSED)
                        zfs_post_state_change(spa, vd, save_state);
```

Alternately, we could perhaps add the event to the end of `spa_vdev_attach()` which handles the replacement itself.

Reply:
OK. I'll look into the second option since that seems simpler to me.

Can you please provide the steps you did to see the duplicate events? Since I didn't see that in my testing, I would like to update my test steps so I don't miss it again.

Should I close this PR and open another one when I have a different patch or just force push this one with the new patch when I have it ready?

Reply:
@cvoltz @tonyhutter I think we should take a step and reconsider the desired behavior.

Currently the code intentionally suppresses all `statechange` events when opening a device.  This is why there are no such events posted when running `zpool import`, when using `zpool add|attach`, or `zpool replace` as in this case.  Instead there are explicit `add`, `attach`, etc, events.

Would be more convenient if we generated events every time we opened a device (offline -> online)?  Our would that generate too much noise in the event logs?

Or should we only generate the events for new devices?  In which case we'd want to handle `zpool create`. `zpool add`, `zpool attach`, `zpool split` as well.  This would in part be redudant with the `config_sync` event which is posted every time there's a configuration change.  This does include a `zpool replace`.

> Can you please provide the steps you did to see the duplicate events?

I was able to reproduce this by running `zpool replace` with the device online.  Two state changes events are posted and you can see one is for the top-level vdev by checking the reported guid in the event.  You can use `zpool status -g` to print each vdevs guid.

> Should I close this PR and open another one when I have a different patch or just force push this one with the new patch when I have it ready?

Let's leave this open at least until we hash out the best way to handle this.  Then I'd suggest force updating the PR so we can keep the discussion in one place.